logind: use pidfd to refer to session leader

msekletar opened the pull request.

msekletar commented:
Another piece in the pidfdize eveything puzzle.

@bluca @poettering I'd like to get your thought on API and approach before adding docs, tests...

poettering commented on src/login/logind-dbus.c:
```diff
@@ -676,36 +676,60 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
 
-static int method_create_session(sd_bus_message *message, void *userdata, sd_bus_error *error) {
-        const char *service, *type, *class, *cseat, *tty, *display, *remote_user, *remote_host, *desktop;
+static int create_session(sd_bus_message *message,
```
we usually add a line break after the `(` when we line break fnction parameters

poettering commented on src/login/logind-dbus.c:
```diff
@@ -676,36 +676,60 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
                 return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid UID");
-        if (leader < 0 || leader == 1 || leader == getpid_cached())
+
+        if (pidfd > 0) {
```
0 is a valid fd. hence this really should check pidfd >= 0 here

poettering commented on src/login/logind-dbus.c:
```diff
@@ -676,36 +676,60 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
+                r = sd_bus_creds_get_pid(creds, (pid_t*) &leader);
+                if (r < 0)
+                        return r;
+        } else if (pid > 0)
```
this will always hold. hence change this to:

```c
} else {
        assert(pid > 0);
        leader = pid;
}
```

this encodes expectations, and acts as a comment of sorts, too

poettering commented on src/login/logind-dbus.c:
```diff
@@ -893,6 +905,8 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
         if (r < 0)
                 goto fail;
 
+        session->leader_pidfd = dup(pidfd);
```
dup() is not ok in our codebase. use `fcntl(fd, FD_DUPFD_CLOEXEC, 3)` instead.

also, check if pidfd >= 0 first, and set to -EBADF otherwise

msekletar replied:
I actually got rid of this after migration to PidRef.

poettering commented on src/login/logind-dbus.c:
```diff
@@ -983,6 +997,49 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
 }
 
+
+
```
spurious three empty lines

poettering commented on src/login/logind-dbus.c:
```diff
@@ -983,6 +997,49 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
+                return r;
+
+        return create_session(message, userdata, error,
+                              uid, -1, leaderfd, service, type, class, desktop, cseat, vtnr, tty, display, remote, remote_user, remote_host);
```
we usually use 0 as invalid pid. should really do here too. given you specify the leaderfd  here that should be fine

bluca replied:
`/* pid= */ 0`

poettering commented on src/login/logind-session.c:
```diff
@@ -935,6 +935,9 @@ int session_finalize(Session *s) {
         }
 
+        if (s->leader_pidfd > 0)
+                close(s->leader_pidfd);
```
drop the check and use safe_close()

poettering replied:
also invalidate the variable, i.e.:

```c
s->leader_pidfd = safe_close(s->leader_pidfd);
``

poettering commented on src/login/logind-session.c:
```diff
@@ -1232,6 +1235,10 @@ bool session_may_gc(Session *s, bool drop_not_started) {
                         return false;
         }
 
+        if (s->leader_pidfd > 0 && s->leader > 0)
```
>= 0

poettering commented on src/login/logind-session.c:
```diff
@@ -1232,6 +1235,10 @@ bool session_may_gc(Session *s, bool drop_not_started) {
         }
 
+        if (s->leader_pidfd > 0 && s->leader > 0)
+                if (pid_is_alive(s->leader))
```
please add pidfd_is_alive() that first translates pidfd to pid internally, and then does the pid_is_alive() check via the pid, and then checks if the pid still matches the pidfd afterwards, before returning the result. So that the pid cannot be reused while we operate on it

poettering commented on src/login/pam_systemd.c:
```diff
@@ -945,6 +946,10 @@ _public_ PAM_EXTERN int pam_sm_open_session(
         if (!IN_SET(r, PAM_SUCCESS, PAM_NO_MODULE_DATA))
                 return pam_syslog_pam_error(handle, LOG_ERR, r, "Failed to get PAM systemd.runtime_max_sec data: @PAMERR@");
 
+        pidfd = pidfd_open(getpid_cached(), 0);
```
should have fallback in case of old kernels.

poettering commented:
so this looks OK, but is of course incomplete: logind passes the "leader" pid on to pid1 via the scope concept. But pid1 doesn't track scope processes via pidfd yet. Which is something we really should add.

So this is still racy, though the race is certainly smaller now. And I think it's OK to merge even if the fix in PID 1 is not coming soon.

bluca commented on src/login/logind-dbus.c:
```diff
@@ -983,6 +997,49 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
+                return r;
+
+        return create_session(message, userdata, error,
+                              uid, leader, -EBADF, service, type, class, desktop, cseat, vtnr, tty, display, remote, remote_user, remote_host);
```
`/* pidfd= */ -EBADF`

bluca commented on src/login/logind-dbus.c:
```diff
@@ -3440,6 +3497,30 @@ static const sd_bus_vtable manager_vtable[] = {
                                               "b", existing),
                                 method_create_session,
                                 0),
+        SD_BUS_METHOD_WITH_ARGS("CreateSessionWithPIDFd",
```
other D-Bus methods we have use `PIDFD` instead of `PIDFd`, so let's keep that naming convention to be consistent

bluca commented on src/login/logind-dbus.c:
```diff
@@ -3440,6 +3497,30 @@ static const sd_bus_vtable manager_vtable[] = {
+                                            "b", remote,
+                                            "s", remote_user,
+                                            "s", remote_host,
+                                            "a(sv)", properties),
```
when adding new complex D-Bus methods nowadays I like to always also append a `t flags` argument, even if unused at the beginning. Often enough we find that we want to do slight tweak or optional changes, and it turns out to be useful.

bluca commented:
> Another piece in the pidfdize eveything puzzle.
> 
> @bluca @poettering I'd like to get your thought on API and approach before adding docs, tests...

Yeah I think this is a great starting point and we should add this interface even if the story is not complete end-to-end

msekletar commented on src/login/logind-session-dbus.c:
```diff
@@ -826,9 +835,25 @@ int session_send_create_reply(Session *s, sd_bus_error *error) {
-        if (fifo_fd < 0)
-                return fifo_fd;
+        if (s->leader_pidfd < 0) {
+                fifo_fd = session_create_fifo(s);
```
Hmm, but why keep it if we don't send it over...maybe we could send it but it is a tiny waste of resources to "leak" the FD we don't really need anymore

poettering replied:
the resources for one fifo fd really don't matter... let's just continue to return that, please.

msekletar replied:
I brought FIFO back.

poettering commented on src/basic/process-util.c:
```diff
@@ -1592,6 +1592,19 @@ int pidfd_verify_pid(int pidfd, pid_t pid) {
+
+        r = pid_is_alive(pid);
+
+        return pidfd_verify_pid(pidfd, pid) < 0 ? -ESRCH : r;
```
please propagate the exact error code pidfd_verify_pid() returns, if it fails.

ret = pid_is_alive(pid);

r = pidfd_verify_pid(pidfd, pid);
if (r < 0)
        return r;

return ret;

poettering commented on src/login/logind-dbus.c:
```diff
@@ -893,6 +907,13 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
+        if (pidfd >= 0) {
+                session->leader_pidfd = fcntl(pidfd, F_DUPFD_CLOEXEC, 3);
+                if (session->leader_pidfd < 0)
+                        goto fail;
```
`r = -errno` is missing here

poettering commented on src/login/logind-session-dbus.c:
```diff
@@ -837,27 +862,54 @@ int session_send_create_reply(Session *s, sd_bus_error *error) {
                         false);
+
+
+
```
spurious empty lines

poettering commented on src/login/logind-session.c:
```diff
@@ -1232,6 +1234,10 @@ bool session_may_gc(Session *s, bool drop_not_started) {
         }
 
+        if (s->leader_pidfd >= 0)
+                if (pidfd_is_alive(s->leader_pidfd))
```
in new code we really try to at least debug log about unexpected errors, hence maybe:

```c
if (s->leader_pidfd >= 0) {
        r = pidfd_is_alive(s->leader_pidfd);
        if (r < 0)
                log_debug_errno(r, "Unable to determine if leader PID " PID_FMT " is still alive, assuming not.");
        if (r > 0)
                return false;
```

poettering commented on src/login/pam_systemd.c:
```diff
@@ -945,6 +946,10 @@ _public_ PAM_EXTERN int pam_sm_open_session(
                 return pam_syslog_pam_error(handle, LOG_ERR, r, "Failed to get PAM systemd.runtime_max_sec data: @PAMERR@");
 
+        pidfd = pidfd_open(getpid_cached(), 0);
+        if (pidfd < 0 && errno != ENOSYS)
```
also, please debug log in the compat case

poettering commented:
Now that #29142 has been merged this should be reworked around the new "PidRef" structure.

poettering commented on src/login/logind-session-dbus.c:
```diff
@@ -837,27 +862,54 @@ int session_send_create_reply(Session *s, sd_bus_error *error) {
+                          s->seat ? s->seat->id : "",
+                          (uint32_t) s->vtnr);
+
+
```
spurious double empty line

YHNdnzj commented on src/basic/process-util.c:
```diff
@@ -1597,6 +1597,23 @@ int pidfd_verify_pid(int pidfd, pid_t pid) {
+
+        ret = pid_is_alive(pid);
+
+        r = pidfd_verify_pid(pidfd, pid);
```
This is pointless. In `pidfd_verify_pid` we ask the pidfd for the pid to compare, which is where the pid originates from...

msekletar replied:
@poettering explicitly requested this behavior I think ...

> please add pidfd_is_alive() that first translates pidfd to pid internally, and then does the pid_is_alive() check via the pid, and then checks if the pid still matches the pidfd afterwards, before returning the result. So that the pid cannot be reused while we operate on it

YHNdnzj replied:
Oops, sorry. Ignore me then.

YHNdnzj replied:
Hmm, maybe some explanatory comments could be added?

msekletar replied:
I've added the comment

msekletar commented:
@poettering @YHNdnzj @bluca updated, PTAL.

bluca commented on man/org.freedesktop.login1.xml:
```diff
@@ -80,6 +80,30 @@ node /org/freedesktop/login1 {
                     out u vtnr,
                     out b existing);
       @org.freedesktop.systemd1.Privileged("true")
+      CreateSessionWithPIDFD(in  u uid,
```
list this in the 'history' section at the bottom of the doc

bluca commented on src/login/logind-dbus.c:
```diff
@@ -984,6 +1002,48 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
+                return r;
+
+        return create_session(message, userdata, error,
+                              uid, leader, -EBADF /* pidfd= */, service, type, class, desktop, cseat, vtnr, tty, display, remote, remote_user, remote_host, 0 /* flags */);
```
comment with the var name is usually before the value, also probably need more line breaks, usually one param per line

msekletar replied:
When it comes to line breaks I think it would make sense to make this consistent with `sd_bus_message_read()` above.

poettering replied:
well, that sd_bus_message_read() line kinda the outlier here...

bluca commented on src/login/logind-dbus.c:
```diff
@@ -984,6 +1002,48 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
+                return r;
+
+        return create_session(message, userdata, error,
+                              uid, 0 /* pid */, leaderfd, service, type, class, desktop, cseat, vtnr, tty, display, remote, remote_user, remote_host, flags);
```
as above

bluca commented on src/login/logind-session.c:
```diff
@@ -529,7 +535,7 @@ int session_load(Session *s) {
                         log_debug_errno(r, "Failed to parse leader PID of session: %s", leader);
                 else {
-                        r = session_set_leader(s, pid);
+                        r = session_set_leader(s, pid, -EBADF /* pidfd */);
```
as above

poettering commented on src/login/logind-session-dbus.c:
```diff
@@ -830,6 +839,22 @@ int session_send_create_reply(Session *s, sd_bus_error *error) {
                 return fifo_fd;
 
+        if (s->leader.fd >= 0) {
+                int r;
```
`int r` as our go-to error code variable always gets the privilege to be decalred on outermost function scope. Hence pleasemove this up. And then merge the two if conditions so there's only one block.

poettering commented on src/login/logind-session-dbus.c:
```diff
@@ -830,6 +839,22 @@ int session_send_create_reply(Session *s, sd_bus_error *error) {
+
+                        (void) sd_event_source_set_description(s->leader_pidfd_event_source, "session-pidfd");
+                }
+        }
```
please move this logic to a new helper session_watch_pidfd() that is part of logind-session.c

poettering commented on src/login/logind-session.c:
```diff
@@ -182,9 +182,15 @@ int session_set_leader(Session *s, pid_t pid) {
+                r = pidref_set_pid(&pidref, pid);
+                if (r < 0)
+                        return r;
+        }
```
please move the `if (r < 0) return r;` parts out of the inner blocks so that you can share it between both blocks.

poettering commented on src/login/logind-session.c:
```diff
@@ -945,6 +953,8 @@ int session_finalize(Session *s) {
                 seat_save(s->seat);
         }
 
+        s->leader.fd = safe_close(s->leader.fd);
```
pidref_done

poettering commented on src/login/logind-session.c:
```diff
@@ -1242,6 +1252,14 @@ bool session_may_gc(Session *s, bool drop_not_started) {
         }
 
+        if (s->leader.fd >= 0) {
+                r = pidfd_is_alive(s->leader.fd);
```
so I'd like you to add a pidref_is_alive() wrapper on top of your new pidfd_is_alive() that uses pidfd_is_alive() if we have a pidfd, and otherwise uses pid_is_alive().

Then just call this here, generically.

poettering commented on src/login/pam_systemd.c:
```diff
@@ -945,6 +946,10 @@ _public_ PAM_EXTERN int pam_sm_open_session(
                 return pam_syslog_pam_error(handle, LOG_ERR, r, "Failed to get PAM systemd.runtime_max_sec data: @PAMERR@");
 
+        pidfd = pidfd_open(getpid_cached(), 0);
+        if (pidfd < 0 && !ERRNO_IS_NOT_SUPPORTED(r))
```
errno, not r

msekletar replied:
Good catch!

msekletar commented:
Updated.

DaanDeMeyer reviewed (approved):
Two small nits

YHNdnzj commented on src/login/logind-dbus.c:
```diff
@@ -677,36 +677,66 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
+        if (flags != 0)
+                return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "Flags must be zero.");
+
+        if (pidfd >= 0) {
+                r = pidfd_get_pid(pidfd, &leader);
+                if (r < 0)
+                        return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "Failed to get PID from PID fd");
+        } else if (pid == 0) {
+                _cleanup_(sd_bus_creds_unrefp) sd_bus_creds *creds = NULL;
+
+                r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_PID, &creds);
+                if (r < 0)
+                        return r;
+
+                r = sd_bus_creds_get_pid(creds, &leader);
+                if (r < 0)
+                        return r;
+        } else  {
+                assert(pid > 0);
+                leader = pid;
+        }
```
I.e. operate on a `PidRef` here.

poettering commented on src/login/logind-dbus.c:
```diff
@@ -677,36 +677,66 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
-        uid_t uid;
-        int remote;
-        uint32_t vtnr = 0;
+        pid_t leader = 0;
```
Please just make this a PidRef right-away

poettering commented on src/login/logind-dbus.c:
```diff
@@ -677,36 +677,66 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
+                return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "Flags must be zero.");
+
+        if (pidfd >= 0) {
+                r = pidfd_get_pid(pidfd, &leader);
```
please make `leader` a PidRef right away, so that you then can use pidref_set_pid() here

poettering commented on src/login/logind-dbus.c:
```diff
@@ -677,36 +677,66 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
+                        return r;
+        } else  {
+                assert(pid > 0);
+                leader = pid;
```
and pidref_set_pid() here

poettering commented on src/login/logind-dbus.c:
```diff
@@ -890,7 +908,7 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
 
         session_set_user(session, user);
-        r = session_set_leader(session, leader);
+        r = session_set_leader(session, leader, pidfd);
```
change session_set_leader() to take a PidRef please.

msekletar commented on src/basic/pidref.c:
```diff
@@ -260,6 +260,13 @@ int pidref_verify(PidRef *pidref) {
 }
 
+
+int pidref_is_alive(PidRef *pidref) {
```
Hmm, I find this difficult to implement. `pidref_is_alive()` requires prototypes of `pidfd_is_alive()` and `pid_is_alive()`. However, if I include `process-util.h` in `pidref.h` I am getting include loop. If I define these prototypes manually then I am getting a ton of warnings because we compile with `-Wredundant-decls`.

poettering replied:
don't sweat it I'd say ;-)

poettering replied:
(but I'd put some more work on you, sorry: if you add pidref_is_alive() please move relevant candidates over to using it. Specifically in src/core/service.c there are two invocations if pid_is_alive() you can trivially switch over to pidref_is_alive(). once you did that, please remove the line that says `- pid_is_alive() → pidref_is_alive()` in TODO, since you moved one more thing over, successfully).

msekletar replied:
Will do.

msekletar commented:
@poettering @YHNdnzj updated. Thank you for previous rounds, hopefully we are close to merge now...but let's see :)

YHNdnzj commented on src/basic/pidref.c:
```diff
@@ -260,6 +260,12 @@ int pidref_verify(PidRef *pidref) {
         return 1; /* We have a pidfd and it still points to the PID we have, hence all is *really* OK → return 1 */
 }
 
+int pidref_is_alive(PidRef *pidref) {
```
Make `*pidref` const?

YHNdnzj commented on src/basic/pidref.h:
```diff
@@ -56,6 +56,7 @@ int pidref_kill_and_sigcont(PidRef *pidref, int sig);
 int pidref_sigqueue(PidRef *pidfref, int sig, int value);
 
 int pidref_verify(PidRef *pidref);
+int pidref_is_alive(PidRef *pidref);
```
(while you're at it, maybe change `pidref_verify` to take const PidRef* too)

YHNdnzj commented on src/basic/pidref.c:
```diff
@@ -260,6 +260,12 @@ int pidref_verify(PidRef *pidref) {
+int pidref_is_alive(PidRef *pidref) {
+        assert(pidref);
+
+        return pidref->fd >= 0 ? pidfd_is_alive(pidref->fd) : pid_is_alive(pidref->pid);
```
Here, if `pidfd_is_alive` is called, return value would be `-ESRCH` if process is already dead, but for `pid_is_alive` `false` is returned.

YHNdnzj commented on src/login/logind-session.c:
```diff
@@ -1237,6 +1268,12 @@ bool session_may_gc(Session *s, bool drop_not_started) {
 
+        r = pidref_is_alive(&s->leader);
+        if (r < 0)
+                log_debug_errno(r, "Unable to determine if leader PID " PID_FMT " is still alive, assuming not.", s->leader.pid);
```
... and thus the log message incorrect, since we might get `-ESRCH` as a valid result.

keszybz replied:
I'm not sure if `pidref_is_alive` return is handled correctly. It will return -ESRCH if the field is unset. Maybe `pidref_is_alive` should be changed to return false instead?

msekletar replied:
I think it is OK to put session in closing state if we can't figure out whether the leader is alive and I think it doesn't really matter what the reason is.

YHNdnzj commented on src/basic/process-util.c:
```diff
@@ -1597,6 +1597,27 @@ int pidfd_verify_pid(int pidfd, pid_t pid) {
+         * liveness status of PID only if pidfd still refers to that same PID. */
+        r = pidfd_verify_pid(pidfd, pid);
+        if (r < 0)
+                return r;
```
Also here

poettering commented on src/basic/pidref.c:
```diff
@@ -260,6 +260,12 @@ int pidref_verify(PidRef *pidref) {
+int pidref_is_alive(PidRef *pidref) {
+        assert(pidref);
+
+        return pidref->fd >= 0 ? pidfd_is_alive(pidref->fd) : pid_is_alive(pidref->pid);
```
oh, i just realized one more thing: pidref_is_alive() should return -ESRCH in case it is invoked with an argument initialized to PIDREF_NULL, i.e. where the pidfd is -EBADF and the pid is zero. But with this code we'd invoke pid_is_alive() with 0 as argument, which internally is mapped to our process ID, and we are of course always alive... 

hence this needs to handle the fd < 0 && pid == 0 case and return false.

msekletar replied:
Hmm, I see the logic in `pid_is_alive()` now. I guess adding simple wrappers is not that simple....In any case, what is the rational for this kind of handling of pid `0` in pid_is_alive()?

poettering replied:
in the simple helpers that operate on processes by pid we systematically interpret pid=0 as "our own process", which is what classic UNIX kill() does too, and setpriority() as well and various other tools. It's simply how UNIX does this, and it's one of the things that make sense.

bluca commented:
@msekletar can you provide a rebase for this? Then it's ready to go - it's an rc blocker

bluca commented:
/packit build

YHNdnzj commented on src/login/logind-session.c:
```diff
@@ -170,23 +170,16 @@ void session_set_user(Session *s, User *u) {
         int r;
 
         assert(s);
+        assert_se(pidref_is_set(&leader));
```
Why `assert_se`?

msekletar replied:
You are right there is no reason for `assert_se()` here. Fixed.

bluca commented on src/login/logind-session.c:
```diff
@@ -170,23 +170,16 @@ void session_set_user(Session *s, User *u) {
         int r;
 
         assert(s);
+        assert(pidref_is_set(&leader));
```
assert is not right here, as it will be removed depending on the compiler options - instead just check and return an error

msekletar replied:
Calling the function w/o leader set is programming error so isn't assert() better option here?

bluca replied:
I might be confused with the other calls, if this one doesn't have side effects and it's just a check then it's fine

bluca commented:
This is related:

```
autopkgtest [16:02:43]: test logind: [-----------------------
 * try-restarting systemd-logind
 * daemon is started
 * loginctl succeeds
 * XDG_SESSION_ID=
FAIL: process is not in session cgroup
/proc/self/cgroup:
13:hugetlb:/
12:devices:/
11:misc:/
10:freezer:/
9:blkio:/
8:memory:/
7:pids:/
6:rdma:/
5:cpuset:/
4:perf_event:/
3:cpu,cpuacct:/
2:net_cls,net_prio:/
1:name=systemd:/
0::/.lxc
No sessions.
Failed to get path for session '': Caller does not belong to any known session.
```

This is the test case:

https://salsa.debian.org/systemd-team/systemd/-/blob/upstream-ci/debian/tests/logind?ref_type=heads

msekletar commented:
@bluca I think I know what is going on. The problem seems to be that when we start the `test-logind` we have already upgraded systemd package and thus `pam_systemd.so` is already new, i.e. will call `CreateSessionWithPIDFD()` but systemd-logind wasn't restarted yet. We restart systemd-logind as part of the test but we failed to create the session. Later we check if we are session member but that is not the case because `CreateSessionWithPIDFD()` failed.

I think to fix this we can implement fallback in `pam_systemd.so` so that we try to call legacy `CreateSession()` API if `CreateSessionWithPIDFD()` is not available. Alternatively we could restart systemd-logind on package upgrades.

@poettering what do you think?

poettering commented on src/login/logind-session-dbus.c:
```diff
@@ -830,6 +831,12 @@ int session_send_create_reply(Session *s, sd_bus_error *error) {
         if (fifo_fd < 0)
                 return fifo_fd;
 
+        if (s->leader.fd >= 0) {
```
please don't do this test here. instead just call session_watch_pidfd() and make it exit early if we don't have the pidfd. i.e. abstract the test away.

poettering commented on src/login/logind-session.c:
```diff
@@ -170,23 +170,16 @@ void session_set_user(Session *s, User *u) {
-                return -EINVAL;
-
-        if (s->leader.pid == pid)
+        if (s->leader.pid == leader.pid)
```
use pidref_equal()

poettering commented on src/login/logind-dbus.c:
```diff
@@ -890,7 +915,7 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
 
         session_set_user(session, user);
-        r = session_set_leader(session, leader);
+        r = session_set_leader(session, TAKE_PIDREF(leader));
```
so here you donate the PidRef to session_set_leader() even on error, but the function is not ready for that.

msekletar replied:
Thank you! This is good catch indeed!

poettering commented on src/login/logind-session.c:
```diff
@@ -170,23 +170,16 @@ void session_set_user(Session *s, User *u) {
-
-        if (s->leader.pid == pid)
+        if (s->leader.pid == leader.pid)
                 return 0;
```
memory mgmt is borked here now. you donated the PidRef to session_set_leader() in the earlier call, but here you don't release it if you exit early.

you basically have three options here:

1. donate the pidref to this function in both error and success. in that case i'd immediately assign the parameter to a local variable that uses `_cleanup_` to release the thing if the scope is left. i.e. rename the param `_pidref` and then do `_cleanup_(pidref_done) PidRef pidref = _pidref;` or so.

2. donate it only on success, in which case you make sure that if you exit >= 0 you use TAKE_PIDREF() once, but on the failure paths you just leave the thing as it is.

3. you make a clone of the passed fd, i.e. no dontation takes place. in that case you use pidref_set_pidfd() and pass the pidref.fd of the parameter to it.

Usually when we go for option 1 we indicate that in the function name, i.e. rename it to xyz_consume() or so, since our typical rule is to take possession of things only on success, not on failure.

poettering commented on src/login/logind-session.c:
```diff
@@ -529,7 +522,13 @@ int session_load(Session *s) {
-                        r = session_set_leader(s, pid);
+                        _cleanup_(pidref_done) PidRef p = PIDREF_NULL;
+
+                        r = pidref_set_pid(&p, pid);
```
instead of parse_pid() + pidref_set_pid() you can also use pidref_set_pidstr() which does this in one call

poettering commented on src/login/logind-session.c:
```diff
@@ -1226,6 +1229,34 @@ static void session_remove_fifo(Session *s) {
+        int r;
+
+        assert(s);
+        assert(s->leader.fd >= 0);
```
as mentioned earlier, i think it would be good if this function would just do an early clean exit if we have no pidfd. i.e. change this assert() here to an `if (!…) return 0`

poettering commented on src/login/logind-session.c:
```diff
@@ -1282,7 +1319,7 @@ SessionState session_get_state(Session *s) {
                 return SESSION_CLOSING;
 
-        if (s->scope_job || s->fifo_fd < 0)
+        if (s->scope_job || (s->leader.fd < 0 && s->fifo_fd < 0))
```
!pidref_is_set()

msekletar commented:
@poettering Updated. I fixed everything mentioned above and I've also added fallback in `pam_systemd` but I am not sure about the approach taken (helper struct in order to avoid passing a lot of parameters around). PTAL.

bluca commented on src/login/pam_systemd.c:
```diff
@@ -945,6 +1042,7 @@ _public_ PAM_EXTERN int pam_sm_open_session(
         if (!IN_SET(r, PAM_SUCCESS, PAM_NO_MODULE_DATA))
                 return pam_syslog_pam_error(handle, LOG_ERR, r, "Failed to get PAM systemd.runtime_max_sec data: @PAMERR@");
 
+
```
spurious newline

bluca commented:
```
../src/login/pam_systemd.c:903:29: error: unused variable 'pidfd' [-Werror,-Wunused-variable]
        _cleanup_close_ int pidfd = -EBADF;
                            ^
```

keszybz commented on man/org.freedesktop.login1.xml:
```diff
@@ -525,10 +551,10 @@ node /org/freedesktop/login1 {
-      open or close login sessions. These calls should <emphasis>never</emphasis> be invoked directly by
-      clients. Creating/closing sessions is exclusively the job of PAM and its
-      <citerefentry><refentrytitle>pam_systemd</refentrytitle><manvolnum>8</manvolnum></citerefentry>
+      <para><function>CreateSession()</function>, <function>CreateSessionWithPIDFD()</function> and
```
Oxford comma please.

msekletar replied:
Fixed.

keszybz reviewed (commented):
LGTM.
